src/stitcher/image_loader.rs

```diff
@@ -1,26 +1,65 @@
 //! This module is for all methods involved in getting selected images loaded into memory.
 
-use anyhow::anyhow;
 use image::{
-    image_dimensions, imageops::FilterType::Lanczos3, GenericImage, ImageReader, RgbImage,
+    error::ImageError, image_dimensions, imageops::FilterType::Lanczos3, GenericImage, ImageReader,
+    RgbImage,
 };
 use rayon::iter::{IntoParallelIterator, ParallelIterator};
-use std::{fs::read_dir, path::PathBuf, time::Instant};
+use std::{
+    fs::read_dir,
+    path::{Path, PathBuf},
+    time::Instant,
+};
 use thiserror::Error;
 
 #[derive(Error, Debug)]
+/// Errors related to loading images.
 pub enum ImageLoaderError {
+    #[error("Expected a directory at \"{0}\"")]
+    ExpectedDirectory(PathBuf),
+    #[error("Could not find a valid path at \"{0}\"")]
+    NotFound(PathBuf),
+    #[error("Permission denied while attempting to access \"{0}\"")]
+    PermissionDenied(PathBuf),
     #[error("No images were found in the selected directory")]
     NoImagesInDirectory,
+    // TODO: convert upstream errors to more specific errors
+    #[error("{0}")]
+    ImageError(ImageError),
+}
+
+impl From<ImageError> for ImageLoaderError {
+    fn from(value: ImageError) -> Self {
+        Self::ImageError(value)
+    }
+}
+
+impl ImageLoaderError {
+    fn from_io_error(err: std::io::Error, path: PathBuf) -> ImageLoaderError {
+        use std::io::ErrorKind as Kind;
+        match err.kind() {
+            Kind::NotFound => ImageLoaderError::NotFound(path),
+            Kind::PermissionDenied => ImageLoaderError::PermissionDenied(path),
+            _ => unimplemented!(),
+        }
+    }
 }
 
 /// Finds all `.jpg`, `.jpeg`, `.png` and `.webp` images within a directory.
 ///
 /// Throws an error if:
 ///  - The directory is invalid or does not contain any images.
 ///  - The directory does not contain any jpg, jpeg, png, or webp images.
-pub fn find_images(directory_path: &str) -> anyhow::Result<Vec<PathBuf>> {
-    let mut images: Vec<_> = read_dir(directory_path)?
+pub fn find_images(directory_path: impl AsRef<Path>) -> Result<Vec<PathBuf>, ImageLoaderError> {
+    // create pathbuf, check if path is a directory
+    let path = PathBuf::from(directory_path.as_ref());
+    if !path.is_dir() {
+        return Err(ImageLoaderError::ExpectedDirectory(path));
+    }
+
+    // get images
+    let mut images: Vec<_> = read_dir(directory_path)
+        .map_err(|e| ImageLoaderError::from_io_error(e, path))?
```

Comment:
All the `map_err` usage is required because `ImageLoaderError` holds additional context (the path in question) that `io::Error` doesn't provide, so a `From<io::Error` implementation isn't possible.